### DISCUSSION TITLE
java.lang.NoClassDefFoundError: scala/collection/immutable/Seq

### DISCUSSION THREAD

**Discussion by danielbasso** [scalameta/metals#3342] [Q&A] [answered]
Hello,

I received a Gradle Scala project from a colleague that is working in production, and I'm trying to run it on my machine. When I try to run, this error appear:

```
Exception in thread "main" java.lang.NoClassDefFoundError: scala/Serializable
    (...)
Caused by: java.lang.ClassNotFoundException: scala.Serializable
```

I can't find a good answer to this error online. I also should mention that I receive this message when running from Visual Studio Code:

`2021.12.03 18:11:49 INFO  no build target: using presentation compiler with only scala-library: 2.13.5`

Gradle settings:

```
    apply plugin: 'scala'
    apply plugin: 'java'
    apply plugin: 'idea'

    idea {
    	module {
    		inheritOutputDirs = true
    		downloadJavadoc = false
    		downloadSources = false
    	}
    }

    sourceCompatibility = 1.11
    targetCompatibility = 1.11
    def scalaVersion = '2.13.5'
    def scalaRelease = scalaVersion.split('\\.').take(2).join('.')

    repositories {
    	mavenCentral()
    	maven {url "https://repository.cloudera.com/artifactory/cloudera-repos/"}
    }

    //def deploy = 'desm'
    def deploy = 'prod'

    dependencies {
    	if (deploy == 'prod') {
    		compileOnly "org.scala-lang:scala-library:$scalaVersion"
    		compileOnly group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.11.2'
    		compileOnly group: 'com.fasterxml.jackson.module', name: "jackson-module-scala_$scalaRelease", version: '2.11.2'
    		compileOnly group: 'org.apache.hbase', name: 'hbase-client', version: '2.1.0-cdh6.3.3'
    		compileOnly group: 'org.apache.kafka', name: 'kafka-clients', version: '2.2.1-cdh6.3.3'
    	}else if (deploy == 'desm'){
    		compile "org.scala-lang:scala-library:$scalaVersion"
    		compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.11.2'
    		compile group: 'com.fasterxml.jackson.module', name: "jackson-module-scala_$scalaRelease", version: '2.11.2'
    		compile group: 'org.apache.hbase', name: 'hbase-client', version: '2.1.0-cdh6.3.3'
    		compile group: 'org.apache.kafka', name: 'kafka-clients', version: '2.2.1-cdh6.3.3'
    	}
    	compile group: 'ml.dmlc', name: 'xgboost4j', version: '0.72', {
    		exclude group: 'com.typesafe.akka', module: 'akka-actor_2.11'
    		exclude group: 'com.typesafe.akka', module: 'akka-testkit_2.11'
    		exclude group: 'junit', module: 'junit'
    		exclude group: 'org.scala-lang', module: 'scala-compiler'
    		exclude group: 'org.scala-lang', module: 'scala-reflect'
    		exclude group: 'org.scala-lang', module: 'scala-library'
    		exclude group: 'org.scalatest',  module: 'scalatest_2.11'
    	}
    }
```

JVM: 11.0.13  
XGBoost: 0.72 (It's an old version, I know)  
Scala: 2.13.5

Some answers online point out to the absence of an library called `scala-library`. In my project, this library does appear:

![image](https://user-images.githubusercontent.com/20959177/144901786-cd4f114b-8159-4fd3-966d-4d16ae68ade2.png)

And in bloop:

![image](https://user-images.githubusercontent.com/20959177/144902109-4593eb06-621a-4fc4-bc30-19c725c45dd6.png)

![image](https://user-images.githubusercontent.com/20959177/144902202-cf6547f3-e68d-47fc-8066-693de4f5a45e.png)

What could be wrong?

**Comment by kpodsiad** [comment, marked as answer]
Looks like https://github.com/scala/bug/issues/11832

**Reply by kpodsiad** [reply]
I wonder why this error isn't happening at production, maybe it's a part of the code which runs only in dev?
You could use [analyze stack trace](https://scalameta.org/metals/blog/page/18/#analyze-stacktrace-command) command in metals to easily navigate to code from the stacktrace.

**Reply by danielbasso** [reply]
I wonder the same, this is bugging me... I've tried Scala 2.12 and Scala 2.11 as well, no dice. With Scala 2.11 the error changes a bit, it becomes "java.lang.NoClassDefFoundError: scala/collection/Seq". I will try analyze stack trace for now.

**Reply by kpodsiad** [reply]
Have you tried to run code via Intellij? There's an option to add a breakpoint in the place where the exception happened or sth similar if I remember correctly. Maybe it'll help you with the debugging.

**Reply by danielbasso** [reply]
I've just realized something. My project has `.bloop` folder and inside it a `.json` file with the name of my project. This json has two properties:

```
{
    "classpath": [(list of jars - 1)],
    (...)
    "platform": {
            "name": "jvm",
            "config": {
                "home": "/usr/lib/jvm/jdk-11.0.13+8",
                "options": [
                    
                ]
            },
            "mainClass": [
                
            ],
            "classpath": [(another list of jars - 2)]
} 
```

It seems that only the jars in the list nr. 2 count as I'm debugging, or simply trying to run my code. But the second list has waaay less jars than the list nr.1 . If I manually add a jar from the first list to the second, the error related to that jar disappears. Could you explain in layman's terms why is this? There is a simple option to copy all jars from the project classpath to the jvm classpath?

**Reply by kpodsiad** [reply]
That's a boomer, I think it should _just_ work, but I'll take a closer look at this tomorrow.

**Reply by danielbasso** [reply]
I've downgraded the scala version to 2.12.15 than copied the project classpath to the jvm classpath and now it's working. But I'm still little puzzled why they aren't the same to begin with.

**Reply by kpodsiad** [reply]
I'm not a Gradle expert but `compileOnly` caught my attention. According to [this article](https://tomgregory.com/gradle-implementation-vs-compile-dependencies/) dependencies added as compile only ones shouldn't appear on the runtime classpath. It matches what you described, that the second classpath (runtime) is much smaller than the first one (compile).
<details>
  <summary>Here's pic from linked article</summary>
<img width="778" alt="Screenshot 2021-12-07 at 07 51 16" src="https://user-images.githubusercontent.com/37124721/144980804-af4a1661-0fbc-4a7d-bfa7-2faa640331f5.png">
</details>

**Reply by danielbasso** [reply]
That's it. Great advice. The code has:

```scala
//def deploy = 'desm'
def deploy = 'prod'
```

So it was using production settings, which only create the "Compile classpath". The word "compile" is the old word for "implementation", so that makes sense. Thanks a lot!

Now, about the version 2.13.5 I will still have to check it out.
